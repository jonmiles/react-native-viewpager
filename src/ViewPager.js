// @flow

import React, { Component } from "react";
import {
  View,
  StyleSheet,
  Animated,
  Dimensions,
  PanResponder
} from "react-native";

const { width, height } = Dimensions.get("screen");

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  page: {
    position: "absolute",
    top: 0,
    left: 0,
    bottom: 0,
    width,
    alignItems: "center",
    justifyContent: "center"
  }
});

type DefaultProps = {
  initialIndex: number
};

type Props = DefaultProps & {
  renderPage: Function,
  pageData: any[]
};

type State = {
  index: number,
  x: number,
  offset: number,
  xTracker: Animated.Value
};

class ViewPager extends Component<DefaultProps, Props, State> {
  props: Props;
  state: State;

  panResponder: *;
  xTracker: Animated.Value;

  static defaultProps: DefaultProps = {
    initialIndex: 0
  };

  constructor(props: Props) {
    super(props);

    const { initialIndex } = props;

    this.xTracker = new Animated.Value(0);
    this.xTracker.addListener(e => this.setState({ x: e.value }));

    this.state = {
      index: initialIndex,
      x: -width * initialIndex,
      offset: 0,
      xTracker: this.xTracker
    };
  }

  componentWillMount() {
    this.panResponder = PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onMoveShouldSetPanResponder: () => true,
      onPanResponderGrant: () => {
        this.xTracker.stopAnimation();
      },
      onPanResponderMove: (evt: *, gestureState: *) => {
        const { dx } = gestureState;
        this.setState({ offset: gestureState.dx });
        this.xTracker.setOffset(gestureState.dx);
      },
      onPanResponderRelease: (evt: *, gestureState: *) => {
        const { vx } = gestureState;
        const { x, offset } = this.state;
        const stopPosition = x + offset;
        const nextPage = this.getNextPage(stopPosition, vx);
        this.snapToPage(nextPage);
      }
    });
  }

  componentWillUnmount() {
    if (this.xTracker) {
      this.xTracker.removeAllListeners();
    }
  }

  getNextPage(x: number, vx: number = 0) {
    const isThrown = vx && (vx < -0.5 || vx > 0.5);

    let index;
    if (isThrown) {
      index =
        vx > 0
          ? Math.abs(Math.ceil(x / width))
          : Math.abs(Math.floor(x / width));
    } else {
      index = Math.abs(Math.round(x / width));
    }

    return Math.min(Math.max(index, 0), this.props.pageData.length - 1);
  }

  snapToPage = (index: number) => {
    this.setState({ index, offset: 0 });

    this.xTracker.flattenOffset();
    Animated.spring(this.xTracker, {
      toValue: -width * index,
      friction: 6
    }).start();
  };

  renderPages(pageIndex: number) {
    const { pageData, renderPage } = this.props;

    return pageData.map((page: *, index: number) => {
      return Math.abs(index - pageIndex) < 2 ? (
        <View key={index} style={[styles.page, { left: width * index }]}>
          {renderPage ? renderPage(page) : null}
        </View>
      ) : null;
    });
  }

  render() {
    const { index, xTracker } = this.state;

    return (
      <View style={styles.container}>
        <Animated.View
          {...this.panResponder.panHandlers}
          style={[styles.container, { left: xTracker }]}
        >
          {this.renderPages(index)}
        </Animated.View>
      </View>
    );
  }
}

export default ViewPager;
