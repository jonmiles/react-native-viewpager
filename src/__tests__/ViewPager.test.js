import React from "react";
import { Text } from "react-native";
import renderer from "react-test-renderer";

jest.mock("Dimensions");
import ViewPager from "../ViewPager";

describe("src/ViewPager", () => {
  const mockData = [{ label: "1" }, { label: "2" }, { label: "3" }];
  const mockRender = d => <Text>{d.label}</Text>;

  it("renders", () => {
    const rendered = renderer
      .create(<ViewPager pageData={mockData} renderPage={mockRender} />)
      .toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
