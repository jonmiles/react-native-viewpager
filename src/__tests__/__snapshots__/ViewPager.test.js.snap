// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/ViewPager renders 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={undefined}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "flex": 1,
        "left": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 100,
          },
          Object {
            "left": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        1
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 100,
          },
          Object {
            "left": 100,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        2
      </Text>
    </View>
  </View>
</View>
`;
